File: 1008_construct_bst_from_preorder/bst_from_preorder.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Time: O(n)
 * Space: O(n)
 */
func bstFromPreorder(preorder []int) *TreeNode {
	stack := []*TreeNode{}
	root := TreeNode{preorder[0], nil, nil}
	stack = append(stack, &root)
	for i := 1; i < len(preorder); i++ {
		curr := preorder[i]
		last := stack[len(stack)-1]
		if curr < stack[len(stack)-1].Val {
			last.Left = &TreeNode{curr, nil, nil}
			stack = append(stack, last.Left)
		} else {
			for len(stack) > 0 && stack[len(stack)-1].Val < curr {
				last = stack[len(stack)-1]
				stack = stack[:len(stack)-1]
			}
			last.Right = &TreeNode{curr, nil, nil}
			stack = append(stack, last.Right)
		}
	}
	return &root
}

func main() {

}

File: 1008_construct_bst_from_preorder/bst_from_preorder.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def bst_from_preorder(preorder: List[int]) -> TreeNode:
    """
    Time: O(n)
    Space: O(n)
    """
    head = TreeNode(preorder[0])
    for i in preorder[1:]:
        insert(head, TreeNode(i))
    return head


def insert(node: TreeNode, new_node: TreeNode):
    if new_node.val > node.val:
        if not node.right:
            node.right = new_node
        else:
            insert(node.right, new_node)
    else:
        if not node.left:
            node.left = new_node
        else:
            insert(node.left, new_node)
